File: frontend/src/api/ralph.ts
/**
 * Ralph AI Runner API client
 */

const API_BASE = '/api';

// ==================== Types ====================

export type JobType = 'planned' | 'feedback';
export type JobStatus = 'pending' | 'running' | 'approved' | 'skipped' | 'needs_fix' | 'failed';
export type RunStatus = 'pending' | 'running' | 'paused' | 'completed' | 'cancelled' | 'failed';

export interface JobSource {
  file: string;
  line?: number;
  pattern?: string;
  update_on_complete: boolean;
}

export interface JobResult {
  summary: string;
  files: string[];
  assumptions: string[];
}

export interface Job {
  id: string;
  type: JobType;
  description: string;
  context: string;
  source?: JobSource;
  status: JobStatus;
  result?: JobResult;
  feedback_for?: string;
  feedback_text?: string;
  previous_job?: {
    description: string;
    summary: string;
    files: string[];
  };
  sdk_session_id?: string;  // SDK session ID for history retrieval
}

export interface ExecutionBatch {
  id: string;
  workspace: string;
  source_description: string;
  model: string;
  auto_approve: boolean;
  created_at: string;
  jobs: Job[];
  // Inherited from parent chat session (name -> enabled selections)
  mcp_servers?: Record<string, boolean>;
  tools?: Record<string, boolean>;
}

export interface RalphRun {
  id: string;
  batch_id: string;
  workspace: string;
  status: RunStatus;
  current_job_index: number;
  auto_approve: boolean;
  started_at?: string;
  completed_at?: string;
  error?: string;
}

export interface RunSummary {
  id: string;
  batch_id: string;
  workspace: string;
  status: RunStatus;
  current_job_index: number;
  total_jobs: number;
  auto_approve: boolean;
  started_at?: string;
  completed_at?: string;
  current_job_description?: string;
}

export interface CreateBatchRequest {
  workspace: string;
  source_description?: string;
  model: string;
  auto_approve?: boolean;
  jobs: Omit<Job, 'status' | 'result'>[];
  // Inherited from parent chat session (name -> enabled selections)
  mcp_servers?: Record<string, boolean>;
  tools?: Record<string, boolean>;
}

export interface StartRunRequest {
  batch_id: string;
  auto_approve?: boolean;
}

export interface FeedbackRequest {
  feedback_text: string;
}

// Job history for viewing agent output
export interface JobHistoryMessage {
  role: 'user' | 'assistant' | 'tool_start' | 'tool_complete';
  content?: string;
  tool?: string;
  result?: string;
}

export interface JobHistory {
  job_id: string;
  sdk_session_id: string | null;
  status: JobStatus;
  result: JobResult | null;
  messages: JobHistoryMessage[];
  raw_events: Record<string, unknown>[];
  error?: string;
}

// ==================== Batches API ====================

export async function createBatch(request: CreateBatchRequest): Promise<ExecutionBatch> {
  const response = await fetch(`${API_BASE}/ralph/batches`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request),
  });
  if (!response.ok) throw new Error('Failed to create batch');
  return response.json();
}

export async function listBatches(): Promise<ExecutionBatch[]> {
  const response = await fetch(`${API_BASE}/ralph/batches`);
  if (!response.ok) throw new Error('Failed to list batches');
  return response.json();
}

export async function getBatch(batchId: string): Promise<ExecutionBatch> {
  const response = await fetch(`${API_BASE}/ralph/batches/${batchId}`);
  if (!response.ok) throw new Error('Failed to get batch');
  return response.json();
}

export async function deleteBatch(batchId: string): Promise<void> {
  const response = await fetch(`${API_BASE}/ralph/batches/${batchId}`, { method: 'DELETE' });
  if (!response.ok) throw new Error('Failed to delete batch');
}

// ==================== Runs API ====================

export async function startRun(request: StartRunRequest): Promise<RalphRun> {
  const response = await fetch(`${API_BASE}/ralph/runs`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request),
  });
  if (!response.ok) throw new Error('Failed to start run');
  return response.json();
}

export async function listRuns(): Promise<RunSummary[]> {
  const response = await fetch(`${API_BASE}/ralph/runs`);
  if (!response.ok) throw new Error('Failed to list runs');
  return response.json();
}

export async function listActiveRuns(): Promise<RunSummary[]> {
  const response = await fetch(`${API_BASE}/ralph/runs/active`);
  if (!response.ok) throw new Error('Failed to list active runs');
  return response.json();
}

export async function getRun(runId: string): Promise<RunSummary> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}`);
  if (!response.ok) throw new Error('Failed to get run');
  return response.json();
}

export async function getRunFull(runId: string): Promise<{ run: RalphRun; batch: ExecutionBatch | null }> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}/full`);
  if (!response.ok) throw new Error('Failed to get run details');
  return response.json();
}

export async function deleteRun(runId: string): Promise<void> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}`, { method: 'DELETE' });
  if (!response.ok) throw new Error('Failed to delete run');
}

// ==================== Run Control API ====================

export async function approveJob(runId: string): Promise<RalphRun> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}/approve`, { method: 'POST' });
  if (!response.ok) throw new Error('Failed to approve job');
  return response.json();
}

export async function skipJob(runId: string): Promise<RalphRun> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}/skip`, { method: 'POST' });
  if (!response.ok) throw new Error('Failed to skip job');
  return response.json();
}

export async function retryJob(runId: string): Promise<RalphRun> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}/retry`, { method: 'POST' });
  if (!response.ok) throw new Error('Failed to retry job');
  return response.json();
}

export async function submitFeedback(runId: string, feedbackText: string): Promise<RalphRun> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}/feedback`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ feedback_text: feedbackText }),
  });
  if (!response.ok) throw new Error('Failed to submit feedback');
  return response.json();
}

export async function setAutoApprove(runId: string, autoApprove: boolean): Promise<RalphRun> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}/auto-approve?auto_approve=${autoApprove}`, { method: 'POST' });
  if (!response.ok) throw new Error('Failed to set auto-approve');
  return response.json();
}

export async function stopRun(runId: string, force: boolean = false): Promise<RalphRun> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}/stop?force=${force}`, { method: 'POST' });
  if (!response.ok) throw new Error('Failed to stop run');
  return response.json();
}

export async function resumeRun(runId: string): Promise<RalphRun> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}/resume`, { method: 'POST' });
  if (!response.ok) throw new Error('Failed to resume run');
  return response.json();
}

// ==================== Job History API ====================

export async function getJobHistory(runId: string, jobId: string): Promise<JobHistory> {
  const response = await fetch(`${API_BASE}/ralph/runs/${runId}/jobs/${jobId}/history`);
  if (!response.ok) throw new Error('Failed to get job history');
  return response.json();
}

// ==================== Streaming ====================

// Job event types from backend
export interface JobStartEvent {
  type: 'job_start';
  job_id: string;
  job_index: number;
  description: string;
}

export interface JobEventData {
  type: 'job_event';
  job_id: string;
  event_type: string;
  data: {
    content?: string;
    delta?: string;
    tool_name?: string;
    tool_call_id?: string;
  };
}

export interface JobCompleteEvent {
  type: 'job_complete';
  job_id: string;
  sdk_session_id?: string;
  summary?: string;
}

export interface JobErrorEvent {
  type: 'job_error';
  job_id: string;
  error: string;
}

export type JobEvent = JobStartEvent | JobEventData | JobCompleteEvent | JobErrorEvent;

export function streamRun(
  runId: string,
  onState: (run: RalphRun, batch: ExecutionBatch | null) => void,
  onComplete: (status: RunStatus) => void,
  onError: (error: string) => void,
  onJobEvent?: (event: JobEvent) => void  // New callback for job-level events
): () => void {
  const eventSource = new EventSource(`/api/ralph/runs/${runId}/stream`);
  
  eventSource.onmessage = (event) => {
    try {
      const data = JSON.parse(event.data);
      
      if (data.type === 'state') {
        onState(data.run, data.batch);
      } else if (data.type === 'complete') {
        onComplete(data.status);
        eventSource.close();
      } else if (data.type === 'error') {
        onError(data.message);
        eventSource.close();
      } else if (['job_start', 'job_event', 'job_complete', 'job_error'].includes(data.type)) {
        // Pass job events to callback
        onJobEvent?.(data as JobEvent);
      }
    } catch (e) {
      console.error('Error parsing SSE event:', e);
    }
  };
  
  eventSource.onerror = () => {
    onError('Connection lost');
    eventSource.close();
  };
  
  // Return cleanup function
  return () => {
    eventSource.close();
  };
}

// ==================== Backwards Compatibility Aliases ====================
// These allow existing code to work during migration

export type TaskType = JobType;
export type TaskStatus = JobStatus;
export type TaskSource = JobSource;
export type TaskResult = JobResult;
export type Task = Job;
export type ExecutionPlan = ExecutionBatch;
export type CreatePlanRequest = CreateBatchRequest;

export const createPlan = createBatch;
export const listPlans = listBatches;
export const getPlan = getBatch;
export const deletePlan = deleteBatch;
export const approveTask = approveJob;
export const skipTask = skipJob;
export const retryTask = retryJob;

File: src/copilot_agent_console/app/routers/ralph.py
"""Ralph AI Runner API router."""

import asyncio
import json
from typing import AsyncGenerator

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse

from copilot_agent_console.app.models.ralph import (
    CreateBatchRequest,
    ExecutionBatch,
    FeedbackRequest,
    RalphRun,
    RunSummary,
    StartRunRequest,
)
from copilot_agent_console.app.services.ralph_service import get_ralph_service

router = APIRouter(prefix="/ralph", tags=["ralph"])


# ==================== Batches ====================

@router.post("/batches", response_model=ExecutionBatch)
async def create_batch(request: CreateBatchRequest) -> ExecutionBatch:
    """Create a new execution batch."""
    service = get_ralph_service()
    return service.create_batch(request)


@router.get("/batches", response_model=list[ExecutionBatch])
async def list_batches() -> list[ExecutionBatch]:
    """List all execution batches."""
    service = get_ralph_service()
    return service.list_batches()


@router.get("/batches/{batch_id}", response_model=ExecutionBatch)
async def get_batch(batch_id: str) -> ExecutionBatch:
    """Get an execution batch by ID."""
    service = get_ralph_service()
    batch = service.get_batch(batch_id)
    if not batch:
        raise HTTPException(status_code=404, detail=f"Batch {batch_id} not found")
    return batch


@router.delete("/batches/{batch_id}")
async def delete_batch(batch_id: str) -> dict:
    """Delete an execution batch."""
    service = get_ralph_service()
    if not service.delete_batch(batch_id):
        raise HTTPException(status_code=404, detail=f"Batch {batch_id} not found")
    return {"status": "deleted"}


# ==================== Runs ====================

@router.post("/runs", response_model=RalphRun)
async def start_run(request: StartRunRequest) -> RalphRun:
    """Start a new Ralph run for a batch."""
    service = get_ralph_service()
    try:
        return await service.start_run(request)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/runs", response_model=list[RunSummary])
async def list_runs() -> list[RunSummary]:
    """List all runs as summaries."""
    service = get_ralph_service()
    return service.list_run_summaries()


@router.get("/runs/active", response_model=list[RunSummary])
async def list_active_runs() -> list[RunSummary]:
    """List all active runs."""
    service = get_ralph_service()
    active_runs = service.get_active_runs()
    summaries = []
    for run in active_runs:
        summary = service.get_run_summary(run.id)
        if summary:
            summaries.append(summary)
    return summaries


@router.get("/runs/{run_id}", response_model=RunSummary)
async def get_run(run_id: str) -> RunSummary:
    """Get a run summary by ID."""
    service = get_ralph_service()
    summary = service.get_run_summary(run_id)
    if not summary:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    return summary


@router.get("/runs/{run_id}/full", response_model=dict)
async def get_run_full(run_id: str) -> dict:
    """Get full run details including batch."""
    service = get_ralph_service()
    run = service.get_run(run_id)
    if not run:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    
    batch = service.get_batch(run.batch_id)
    
    return {
        "run": run.model_dump(),
        "batch": batch.model_dump() if batch else None,
    }


@router.delete("/runs/{run_id}")
async def delete_run(run_id: str) -> dict:
    """Delete a run (must be completed, cancelled, or failed)."""
    service = get_ralph_service()
    if not service.delete_run(run_id):
        raise HTTPException(status_code=400, detail="Cannot delete active run or run not found")
    return {"status": "deleted"}


# ==================== Run Control ====================

@router.post("/runs/{run_id}/approve", response_model=RalphRun)
async def approve_job(run_id: str) -> RalphRun:
    """Approve the current job and continue."""
    service = get_ralph_service()
    run = await service.approve_job(run_id)
    if not run:
        raise HTTPException(status_code=400, detail="Cannot approve job")
    return run


@router.post("/runs/{run_id}/skip", response_model=RalphRun)
async def skip_job(run_id: str) -> RalphRun:
    """Skip the current job and continue."""
    service = get_ralph_service()
    run = await service.skip_job(run_id)
    if not run:
        raise HTTPException(status_code=400, detail="Cannot skip job")
    return run


@router.post("/runs/{run_id}/retry", response_model=RalphRun)
async def retry_job(run_id: str) -> RalphRun:
    """Retry the current job."""
    service = get_ralph_service()
    run = await service.retry_job(run_id)
    if not run:
        raise HTTPException(status_code=400, detail="Cannot retry job")
    return run


@router.post("/runs/{run_id}/feedback", response_model=RalphRun)
async def submit_feedback(run_id: str, request: FeedbackRequest) -> RalphRun:
    """Submit feedback for the current job."""
    service = get_ralph_service()
    run = await service.submit_feedback(run_id, request)
    if not run:
        raise HTTPException(status_code=400, detail="Cannot submit feedback")
    return run


@router.post("/runs/{run_id}/auto-approve", response_model=RalphRun)
async def set_auto_approve(run_id: str, auto_approve: bool) -> RalphRun:
    """Set auto-approve mode for a run."""
    service = get_ralph_service()
    run = await service.set_auto_approve(run_id, auto_approve)
    if not run:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    return run


@router.post("/runs/{run_id}/stop", response_model=RalphRun)
async def stop_run(run_id: str, force: bool = False) -> RalphRun:
    """Stop a run (graceful or forced)."""
    service = get_ralph_service()
    run = await service.stop_run(run_id, force=force)
    if not run:
        raise HTTPException(status_code=400, detail="Cannot stop run")
    return run


@router.post("/runs/{run_id}/resume", response_model=RalphRun)
async def resume_run(run_id: str) -> RalphRun:
    """Resume a paused or cancelled run."""
    service = get_ralph_service()
    run = await service.resume_run(run_id)
    if not run:
        raise HTTPException(status_code=400, detail="Cannot resume run")
    return run


# ==================== Job History ====================

@router.get("/runs/{run_id}/jobs/{job_id}/history")
async def get_job_history(run_id: str, job_id: str) -> dict:
    """Get history for a specific job including SDK events and raw events.
    
    Returns messages and tool executions from the job's SDK session.
    """
    service = get_ralph_service()
    history = await service.get_job_history(run_id, job_id)
    if not history:
        raise HTTPException(status_code=404, detail=f"Job {job_id} not found in run {run_id}")
    return history


# ==================== Streaming ====================

@router.get("/runs/{run_id}/stream")
async def stream_run(run_id: str) -> StreamingResponse:
    """Stream run updates via Server-Sent Events."""
    service = get_ralph_service()
    
    # Verify run exists
    run = service.get_run(run_id)
    if not run:
        raise HTTPException(status_code=404, detail=f"Run {run_id} not found")
    
    async def event_generator() -> AsyncGenerator[str, None]:
        async for event in service.stream_run(run_id):
            yield f"data: {json.dumps(event, default=str)}\n\n"
    
    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
        },
    )
